server/datastore/mysql/software.go

```diff
@@ -659,56 +659,45 @@ func (si *softwareIterator) Next() bool {
 	return si.rows.Next()
 }
 
-func (ds *Datastore) ListSoftwareBySourceIter(
+// AllSoftwareIterator Returns an iterator for the 'software' table, filtering out
+// software entries based on the 'query' param.
+func (ds *Datastore) AllSoftwareIterator(
 	ctx context.Context,
-	sources []string,
+	query fleet.SoftwareIterQueryOptions,
 ) (fleet.SoftwareIterator, error) {
-	if len(sources) == 0 {
-		return nil, errors.New("please provide at least one source")
-	}
-
 	var err error
 	var args []interface{}
 
 	stmt := `SELECT 
-		s.id, 
-		s.name,
-		s.version,
-		s.bundle_identifier,
-		s.release,
-		s.vendor,
-		s.arch,
-		s.source 
-	FROM software s
-	WHERE source IN (?)`
-
-	stmt, args, err = sqlx.In(stmt, sources)
+		s.* ,
+		COALESCE(sc.cpe, '') AS generated_cpe
+	FROM software s 
+	LEFT JOIN software_cpe sc ON (s.id=sc.software_id)`
+	// The rows.Close call is done by the caller once iteration using the
+	// returned fleet.SoftwareIterator is done.
 
-	if err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "error while trying to bind sources")
-	}
+	var conditionals []string
+	arg := map[string]interface{}{}
 
-	rows, err := ds.reader.QueryxContext(ctx, stmt, args...) //nolint:sqlclosecheck
-	if err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "error executing SQL statement")
+	if len(query.ExcludedSources) != 0 {
+		conditionals = append(conditionals, "s.source NOT IN (:excluded_sources)")
+		arg["excluded_sources"] = query.ExcludedSources
 	}
-	return &softwareIterator{rows: rows}, nil
-}
 
-// AllSoftwareWithoutCPEIterator Returns an iterator for the 'software' table, filtering out
-// software entries with CPEs and from the sources included in the 'excludedSources' param.
-func (ds *Datastore) AllSoftwareWithoutCPEIterator(ctx context.Context, excludedSources []string) (fleet.SoftwareIterator, error) {
-	var err error
-	var args []interface{}
+	if len(query.IncludedSources) != 0 {
+		conditionals = append(conditionals, "s.source IN (:included_sources)")
+		arg["included_sources"] = query.IncludedSources
+	}
 
-	stmt := `SELECT s.* FROM software s LEFT JOIN software_cpe sc ON (s.id=sc.software_id) WHERE sc.id IS NULL`
-	// The rows.Close call is done by the caller once iteration using the
-	// returned fleet.SoftwareIterator is done.
-	if excludedSources != nil {
-		stmt += ` AND s.source NOT IN (?)`
-		stmt, args, err = sqlx.In(stmt, excludedSources)
+	if len(conditionals) != 0 {
+		cond := strings.Join(conditionals, " AND ")
+		stmt, args, err = sqlx.Named(stmt+" WHERE "+cond, arg)
 		if err != nil {
-			return nil, ctxerr.Wrap(ctx, err, "loads cpes")
+			return nil, ctxerr.Wrap(ctx, err, "load host software")
```

Comment:
Let's use a different string error (same in the `sqlx.In` below), to help troubleshoot.

Reply:
Done in https://github.com/fleetdm/fleet/pull/10514/commits/8c2e90168b77aaf402e96e4b18c5e5bc2ce018c8